core/src/main/java/org/testcontainers/containers/ContainerDef.java

```diff
@@ -0,0 +1,253 @@
+package org.testcontainers.containers;
+
+import com.github.dockerjava.api.command.CreateContainerCmd;
+import com.github.dockerjava.api.model.Bind;
+import com.github.dockerjava.api.model.ExposedPort;
+import com.github.dockerjava.api.model.HostConfig;
+import com.github.dockerjava.api.model.InternetProtocol;
+import com.github.dockerjava.api.model.PortBinding;
+import com.github.dockerjava.api.model.Ports;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.UnstableAPI;
+import org.testcontainers.containers.wait.strategy.WaitStrategy;
+import org.testcontainers.images.RemoteDockerImage;
+import org.testcontainers.utility.Base58;
+import org.testcontainers.utility.DockerImageName;
+import org.testcontainers.utility.TestcontainersConfiguration;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Getter
```

Comment:
We need to be careful with getters to avoid returning the original (mutable) collections

Reply:
I applied https://github.com/testcontainers/testcontainers-java/pull/7714/commits/f3d78c220977187ca6375242920ce53fa1aeaa4a but due to the current mutable nature of GenericContainer and it's usage in some modules (Kafka, Selenium) will break the existing behavior and third-party implementations.